Source/bmalloc/libpas/src/libpas/pas_mar_registry.h

```diff
@@ -0,0 +1,161 @@
+
+#define PAS_MAR_PAGE_SHIFT 14
+
+typedef struct pas_mar_backtrace_record pas_mar_backtrace_record;
+struct pas_mar_backtrace_record {
+    unsigned num_frames;
+    uint32_t hash;
+    void* backtrace_buffer[PAS_MAR_BACKTRACE_MAX_SIZE];
+};
+
+typedef struct pas_mar_memory_action_record pas_mar_memory_action_record;
+struct pas_mar_memory_action_record {
+    void* address;
+    size_t allocation_size;
+    unsigned backtrace_registry_index;
+    uint32_t backtrace_hash;
+    bool is_allocation;
+};
+
+typedef struct pas_mar_registry pas_mar_registry;
+struct pas_mar_registry {
+    struct pas_mar_backtrace_record backtrace_registry[PAS_MAR_TRACKED_BACKTRACES];
+    struct pas_mar_memory_action_record allocation_record_table[PAS_MAR_TRACKED_ALLOCATIONS];
+    // push to the tail of the FIFO, evict from head
+    unsigned allocation_record_table_head;
+    unsigned allocation_record_table_tail;
+};
+
+struct pas_mar_exported_allocation_record {
+    struct pas_mar_backtrace allocation_trace;
+    struct pas_mar_backtrace deallocation_trace;
+    size_t allocation_size;
+    bool is_valid;
+};
+
+PAS_BEGIN_EXTERN_C;
+
+void* pas_mar_record_allocation(pas_mar_registry*, void* address, size_t allocation_size, unsigned num_stack_frames, void** backtrace);
+void* pas_mar_record_deallocation(pas_mar_registry*, void* address, unsigned num_stack_frames, void** backtrace);
+struct pas_mar_exported_allocation_record pas_mar_get_allocation_record(pas_mar_registry*, void* address);
+
+PAS_END_EXTERN_C;
+
+// Paging helpers
+
+PAS_ALWAYS_INLINE void* pas_mar_canonicalize_address(void* address)
+{
+    return (void*)((uintptr_t)(address) & ((1ull << 48) - 1));
+}
+
+PAS_ALWAYS_INLINE uintptr_t pas_mar_address_to_virtual_page_number(void* address)
+{
+    return (uintptr_t)(pas_mar_canonicalize_address(address)) >> PAS_MAR_PAGE_SHIFT;
+}
+
+// FIFO helpers
+
+PAS_ALWAYS_INLINE unsigned pas_mar_allocation_table_head_index(pas_mar_registry* registry)
+{
+    return registry->allocation_record_table_head % PAS_MAR_TRACKED_ALLOCATIONS;
+}
+
+PAS_ALWAYS_INLINE unsigned pas_mar_allocation_table_tail_index(pas_mar_registry* registry)
+{
+    return registry->allocation_record_table_tail % PAS_MAR_TRACKED_ALLOCATIONS;
+}
+
+PAS_ALWAYS_INLINE bool pas_mar_is_allocation_table_full(pas_mar_registry* registry)
+{
+    return (registry->allocation_record_table_head ^ registry->allocation_record_table_tail) == PAS_MAR_TRACKED_ALLOCATIONS;
+}
+
+PAS_ALWAYS_INLINE void pas_mar_increment_allocation_record_table_head(pas_mar_registry* registry)
+{
+    registry->allocation_record_table_head = registry->allocation_record_table_head + 1;
+    registry->allocation_record_table_head %= MAR_ALLOCATION_RECORD_TABLE_FIFO_MODULUS;
+}
+
+PAS_ALWAYS_INLINE void pas_mar_increment_allocation_record_table_tail(pas_mar_registry* registry)
+{
+    registry->allocation_record_table_tail = registry->allocation_record_table_tail + 1;
```

Comment:
This also seems like it needs to be atomic if we're to be able to access it from multiple threads. Is the table intended to be accessed from within a lock context? That seems simpler to me when compared to retrofitting this as a proper lock-free queue